### ISSUE TITLE
Possible to share code for mean statistics?

### ISSUE THREAD

**Issue by QiuhongAnnaWei** [cselab/G-LED#1]
Thank you for sharing your work and making this repository available! 

I am wondering if you can share the code used to compute and plot mean statistics for the 2D flow over backward-facing step (Fig 4-5 in main paper, Fig 4-11 in the supplementary).

If the code is not available, the mathematical expressions used to compute them will be very helpful too.

**Comment by gaohan1234** [comment]
Dear Anna, thanks for reaching out. Please find the below functions.

import numpy as np
import pdb
from scipy import signal
import scipy
from sklearn.metrics import mean_squared_error



def get_velo(U, Tend=80):
	assert U.shape[0] == 1, 'Only process 1 seq per time'
	Ux = U[0,0,:Tend]
	Uy = U[0,1,:Tend]
	return Ux, Uy

def cal_mean(U, X):
	Umean  = np.mean(U[:,:,X], axis = 0)
	return Umean

def cal_std(U, X):
	Umean  = np.std(U[:,:,X], axis = 0)
	return Umean

def cal_Rstress(Ux, Uy, X):
	Uprime = Ux - Ux.mean(axis=0)
	Vprime = Uy - Uy.mean(axis=0)
	UV     = np.mean(Uprime * Vprime,axis =0)
	return UV[:,X]



def get_Emean(du_dns, fsval, height):
	Elist = []
	T = du_dns.shape[0]
	for t in (range(T)):
		f,E = signal.welch(du_dns[t,height], 
					       fs = fsval, 
						   scaling = 'spectrum') 
		Elist.append(E)
	E_mean = np.asarray(Elist).mean(axis=0)
	return f/f[1], E_mean


def eval_energy_spec_idns(sampled_video, height):
	fsval = 1/512
	# get mean
	u_ml = np.mean(sampled_video, axis = 0)
	# fluctaion
	du_ml  = sampled_video - u_ml
	# different normalized height
	f, E_mean_ml  = get_Emean(du_ml, fsval, height)
	return f[1:75]/f[1], E_mean_ml[1:75]


def eval_spatial_correlation(sampled_video,height):
	"""
	sampled_video (dns_video) : dim 1x4xTimex256x128
	"""
	fsval = 1/512
	# get mean
	u_ml = np.mean(sampled_video, axis = 0)
	# fluctaion
	du_ml  = sampled_video - u_ml
	_, E_mean_ml  = get_Emean(du_ml, fsval, height)
	c_ml  = scipy.fft.ifft(E_mean_ml,axis=0)
	c_ml  = (c_ml.real/c_ml.real[0])[:c_ml.shape[0]//2+1]
	return c_ml

def eval_temporal_correlation(sampled_video,height,width):
	"""
	sampled_video (dns_video) : dim 1x4xTimex256x128
	"""
	# extract information
	# get mean
	u_ml = np.mean(sampled_video[:,height,width], axis = 0)
	# fluctaion
	du_ml      =  sampled_video[:,height,width] - u_ml
	E_mean_ml  =  autocorr(du_ml)
	return E_mean_ml

def autocorr(x):
	result = np.correlate(x, x, mode='full')
	result = result[result.size//2:]
	return result/result.max()


def cal_Error(U, U_truth):
	T = U.shape[0]
	assert T == U_truth.shape[0]
	E = []
	for t in range(T):
		E.append(
			mean_squared_error(U[t].flatten(),U_truth[t].flatten())/mean_squared_error(U_truth[t].flatten(),U_truth[t].flatten()*0)\
		)
	return np.asarray(E)

def cal_Error_norelative(U, U_truth):
	T = U.shape[0]
	assert T == U_truth.shape[0]
	E = []
	for t in range(T):
		E.append(
			mean_squared_error(U[t].flatten(),U_truth[t].flatten())
		)
	return np.asarray(E)

**Comment by QiuhongAnnaWei** [comment]
Thank you for responding so timely and for sharing the code!

I am trying to replicate the results in Fig 4-5 in main paper and Fig 4-11 in the supplementary. Can you share the exact settings used to generate them? Specifically,
1. Which entry of the data is used as the starting snapshot? (Is it 9500, as is start_n in main_diff_eval_bfs.py?)
2. How many steps are predicted and averaged over? (Should test_Nt be 150, as in main_diff_eval_bfs.py?)

Also, in the expression for <u(y)> given in the paper on page 3, there's a mention of z. I'm wondering what would z be in the 2D case?
 
Thank you!
